pallets/runtime/src/bridge.rs

```diff
@@ -418,69 +394,96 @@ decl_module! {
+            nonce: bridge_tx.nonce.clone(),
+            tx_hash: bridge_tx.tx_hash.clone(),
+        };
+        let mut tx_details = Self::bridge_tx_details(&bridge_tx.recipient, &bridge_tx_identifier);
+
+        ensure!(
+            tx_details.status != BridgeTxStatus::Frozen
+                && tx_details.status != BridgeTxStatus::Handled,
+            Error::<T>::ProposalAlreadyHandled
+        );
+
+        if Self::frozen() {
+            // Bridge module frozen. Retry this tx again later.
+            return Self::handle_bridge_tx_later(&bridge_tx, Self::timelock());
+        }
+
+        if Self::issue(&bridge_tx.recipient, &bridge_tx.amount).is_ok() {
+            tx_details.status = BridgeTxStatus::Handled;
+            tx_details.execution_block = <system::Module<T>>::block_number();
+            tx_details.amount = bridge_tx.amount.clone();
+            <BridgeTxDetails<T>>::insert(&bridge_tx.recipient, &bridge_tx_identifier, tx_details);
             Self::deposit_event(RawEvent::Bridged(bridge_tx));
+        } else {
+            // Recipient missing CDD. Retry this tx again later.
+            return Self::handle_bridge_tx_later(&bridge_tx, Self::timelock());
         }
         Ok(())
     }
 
     /// Handles a bridge transaction proposal after `timelock` blocks.
     fn handle_bridge_tx_later(
-        bridge_tx: BridgeTx<T::AccountId, T::Balance>,
+        bridge_tx: &BridgeTx<T::AccountId, T::Balance>,
         timelock: T::BlockNumber,
     ) -> DispatchResult {
+        let mut already_tried = 0;
+        let bridge_tx_identifier = BridgeTxIdentifier {
+            nonce: bridge_tx.nonce.clone(),
+            tx_hash: bridge_tx.tx_hash.clone(),
+        };
+        let mut tx_details = Self::bridge_tx_details(&bridge_tx.recipient, &bridge_tx_identifier);
+        match tx_details.status {
+            BridgeTxStatus::Absent => {
+                tx_details.status = BridgeTxStatus::Timelocked;
+                tx_details.amount = bridge_tx.amount.clone();
+            }
+            BridgeTxStatus::Pending(x) => {
+                tx_details.status = BridgeTxStatus::Pending(x + 1);
+                already_tried = x + 1;
+            }
+            BridgeTxStatus::Timelocked => {
+                tx_details.status = BridgeTxStatus::Pending(1);
+                already_tried = 1;
+            }
+            // Already handled/frozen.
+            _ => {
+                return Err(Error::<T>::ProposalAlreadyHandled.into());
+            }
+        }
+
+        if already_tried > 24 {
+            // Limits the exponential backoff to *almost infinity* (~180 years)
+            already_tried = 24;
+        }
+
         let current_block_number = <system::Module<T>>::block_number();
-        let mut unlock_block_number = current_block_number + timelock;
-        let range = T::BlockRangeForTimelock::get();
-        let max_unlock_block_number = unlock_block_number + range - One::one();
+        let mut unlock_block_number =
```

Comment:
I think we should reign in the heuristics here. We need guarantee the tx is handled ultimately.

Reply:
If a tx fails for the first 15 times, It's unlikely that it's ever going to succeed and I don't think we need to waste our resources in retrying it constantly. That being, any pending tx can be immediately executed by anyone by calling the handle tx fn. For the testnet, I'm happy with this approach and we can nail down our backoff curve when we get near the mainnet.

Reply:
Longer-than-lifetime retry periods don't make sense, I think. Unless we are a Swiss bank that is. ;-)

It would be good to either
* after some retry, signal that the retries are going to stop (which would signal the dapp to switch to a manual retry mode)
* cap the retry period at some maximum duration, maybe not 4.5 days (`u16`) but perhaps 9 or 18 days.

Cc @adamdossa

Reply:
Since it can be manually triggered by anyone, I don't mind the backoff being aggressive. Critical for me is that the accounting in the bridge matches the Ethereum client so that we don't lose POLY in the conversion process. I'm happy for amounts to remain locked indefinitely whilst a user sorts out their CDD status and then manually triggers the transaction (or asks Polymath to do this).